MvvmCross/Binding/Mac/Views/MvxTableViewSource.cs

```diff
@@ -89,24 +88,41 @@ private NSView GetOrCreateViewFor(NSTableView tableView, NSTableColumn tableColu
             return view;
         }
 
-		public override NSView GetViewForItem(NSTableView tableView, NSTableColumn tableColumn, nint row)
+        public override NSView GetViewForItem(NSTableView tableView, NSTableColumn tableColumn, nint row)
         {
-            if (this.ItemsSource == null)
+            if (ItemsSource == null)
                 return null;
 
-            var item = this.ItemsSource.ElementAt((int)row);
-            var view = this.GetOrCreateViewFor(tableView, tableColumn);
+            var item = ItemsSource.ElementAt((int)row);
+            var view = GetOrCreateViewFor(tableView, tableColumn);
 
-            var bindable = view as IMvxDataConsumer;
-            if (bindable != null)
+            if (view is IMvxDataConsumer bindable)
                 bindable.DataContext = item;
 
             return view;
         }
 
         protected virtual void CollectionChangedOnCollectionChanged(object sender, NotifyCollectionChangedEventArgs args)
         {
-            this.TryDoAnimatedChange(args);
+            Action action = () =>
+            {
+                if (!UseAnimations)
+                {
+                    ReloadTableData();
+                }
+                else
+                {
+                    if (TryDoAnimatedChange(args))
+                        return;
+
+                    ReloadTableData();
+                }
+            };
+
+            if (NSThread.IsMain)
+                action();
+            else
+                InvokeOnMainThread(action);
```

Comment:
FYI, I recently found out the hard way that this might result in having the table display the wrong items. I'll give you an example.

Let's say you have a background service returning items as they become available and inserting them into the table (no surprise, we have such a service in our app). The insertion is sorted, so items may be added anywhere in the list. Now let's assume that a first item comes in, inserted at position `0`. An `Add at position 0` collection change event is scheduled. Another item arrives shortly after, which will also insert at position `0` because it precedes the previous item in the ordering. Another `Add at position 0` event. Because the collection changed events are executed on the main thread (or their side effects on the table are, it makes little difference in this case), both events may be executed after the items source has already been updated (this turned out to be very likely in our case). Two add events are successfully executed, but both will read their data context from position `0` in the items source. The table will have 2 items, but both are the same item. It's not a problem for complete table reloads of course, but those may be undesirable for large data lists.

I thought about this for a while and don't really see an easy / intuitive way around this that doesn't require forcing access to the items source to occur on the main thread. Personally I have implemented a custom observable collection which maintains both an "up to date" list and a list as it is currently viewed by the UI thread. Changes to the former are immediate, whereas changes to the latter happen on the UI thread and are synchronized with the collection change events. This way I can make immediate changes on any thread which will eventually be efficiently synchronized with the data table.

Reply:
Please open a new issue linking this PR instead of commenting here (it's really difficult to be up to date with GitHub notifications and you can easily miss anything, specially if it's a comment on an already merged PR).

Looks like you've a very special situation there, and I think you solved it in a very efficient way 👌. But still, I think this PR was an improvement over the previous behavior (it just made apps crash in the past).

If you think your solution may fit into the framework, I would love to see it in a PR!

Reply:
@nmilcoff Yeah you're absolutely right, I just saw this pop up on the go and had to comment it somewhere quickly. I'll see if I can create something; the thing is it's kind of a complex problem and unless you understand why it's needed the solution doesn't make an awful lot of sense. So there will have to be some explaining too ;).